packages/objectloader2/src/operations/database.ts

```diff
@@ -0,0 +1,198 @@
+import BatchingQueue from '../helpers/batchingQueue.js'
+import Queue from '../helpers/queue.js'
+import { ObjectLoaderRuntimeError } from '../types/errors.js'
+import { CustomLogger, Item, isBase } from '../types/types.js'
+import { ensureError, isSafari } from '@speckle/shared'
+import { BaseDatabaseOptions } from './options.js'
+import { ICache } from './interfaces.js'
+
+export default class CacheDatabase implements ICache {
+  private static _databaseName: string = 'speckle-cache'
+  private static _storeName: string = 'objects'
+  private _options: BaseDatabaseOptions
+  private _logger: CustomLogger
+
+  private _cacheDB?: IDBDatabase
+
+  private _writeQueue: BatchingQueue<Item> | undefined
+
+  constructor(logger: CustomLogger, options?: Partial<BaseDatabaseOptions>) {
+    this._logger = logger
+    this._options = {
+      ...{
+        indexedDB: globalThis.indexedDB,
+        batchMaxSize: 500,
+        batchMaxWait: 1000,
+        enableCaching: true
+      },
+      ...options
+    }
+  }
+
+  async write(obj: Item): Promise<void> {
+    if (!this._writeQueue) {
+      if (!(await this.setupCacheDb())) {
+        return
+      }
+      this._writeQueue = new BatchingQueue<Item>(
+        this._options.batchMaxSize,
+        this._options.batchMaxWait,
+        (batch: Item[]) => CacheDatabase.cacheSaveBatch(batch, this._cacheDB!)
+      )
+    }
+    this._writeQueue.add(obj)
+  }
+
+  async finish(): Promise<void> {
+    await this._writeQueue?.finish()
+  }
+
+  private openDatabase(dbName: string, storeName: string): Promise<IDBDatabase> {
+    return new Promise((resolve, reject) => {
+      const request = this._options.indexedDB.open(dbName, 1)
+
+      request.onupgradeneeded = (event) => {
+        const db = (event.target as IDBOpenDBRequest).result
+        if (!db.objectStoreNames.contains(storeName)) {
+          db.createObjectStore(storeName)
+        }
+      }
+
+      request.onsuccess = () => resolve(request.result)
+      request.onerror = (e) =>
+        reject(ensureError(e, 'Failed to open IndexedDB database'))
+    })
+  }
+
+  private supportsCache(): boolean {
+    return !!(this._options.enableCaching && this._options.indexedDB)
+  }
+
+  private async setupCacheDb(): Promise<boolean> {
+    if (this._cacheDB !== undefined && !this.supportsCache()) return false
+
+    // Initialize
+    await this.safariFix()
+    this._cacheDB = await this.openDatabase(
+      CacheDatabase._databaseName,
+      CacheDatabase._storeName
+    )
+    return true
+  }
+  async getItems(
```

Comment:
pretty strange to have a `getItems(): void` function. so it retrieves them but doesn't return them? maybe if its useful, this can be converted to an async generator (cause results come in one at a time) https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/AsyncGenerator#async_generator_iteration